godot-cell/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+        state.borrow_state.increment_mut()?;
+        let count = state.borrow_state.mut_count();
+        let value = state.get_ptr();
+
+        // SAFETY:
+        // `increment_mut` succeeded, therefore any existing mutable references do not alias, and no new
+        // references may be made unless this one is guaranteed not to alias those.
+        //
+        // This is the case because the only way for a new `GdMut` or `GdRef` to be made after this, is for
+        // either this guard to be dropped or `set_non_aliasing` to be called.
+        //
+        // If this guard is dropped, then we dont need to worry.
+        //
+        // If `set_non_aliasing` is called, then either a mutable reference from this guard is passed in.
+        // In which case, we cannot use this guard again until the resulting non-aliasing guard is dropped.
+        //
+        // We cannot pass in a different mutable reference, since `set_non_aliasing` ensures any references
+        // matches the ones this one would return. And only one mutable reference to the same value can exist
+        // since we cannot have any other aliasing mutable references around to pass in.
+        unsafe { Ok(MutGuard::new(&self.get_ref().state, count, value)) }
+    }
+
+    /// Set the current mutable borrow as not aliasing any other references.
+    ///
+    /// Will error if there is no current possibly aliasing mutable borrow, or if there are any shared
+    /// references.
+    pub fn set_non_aliasing<'a, 'b>(
+        self: Pin<&'a Self>,
+        current_ref: &'b mut T,
+    ) -> Result<NonAliasingGuard<'b, T>, Box<dyn Error>>
+    where
+        'a: 'b,
+    {
+        let mut state = self.state.lock().unwrap();
+        let current_ptr = state.get_ptr();
+        let ptr = NonNull::from(current_ref);
+
+        if current_ptr != ptr {
+            // it is likely not unsound for this to happen, but it's unexpected
+            return Err("wrong reference passed in".into());
+        }
+
+        state.borrow_state.set_non_aliasing()?;
+        let old_ptr = state.get_ptr();
+        state.set_ptr(ptr);
+
+        Ok(NonAliasingGuard::new(&self.get_ref().state, old_ptr))
+    }
+
+    /// Returns `true` if there are any mutable or shared references, regardless of whether the mutable
+    /// references are aliasing or not.
+    ///
+    /// In particular this means that it is safe to destroy this cell and the value contained within, as no
+    /// references can exist that can reference this cell.
+    ///
+    /// Keep in mind that in multithreaded code it is still possible for this to return true, and then the
+    /// cell hands out a new borrow before it is destroyed. So we still need to ensure that this cannot
+    /// happen at the same time.
+    pub fn is_currently_bound(self: Pin<&Self>) -> bool {
+        let state = self.state.lock().unwrap();
+
+        state.borrow_state.shared_count() > 0 || state.borrow_state.mut_count() > 0
+    }
```

Comment:
> Keep in mind that in multithreaded code it is still possible for this to return true, and then the cell hands out a new borrow before it is destroyed. So we still need to ensure that this cannot happen at the same time.

Yes, very good observation.

In multithreaded contexts, any follow-up actions need to be protected by the mutex itself, which would likely lead to some API like `self.if_currently_bound(|| action)`. Alternatively, surrounding mutexes (e.g. storage) could take over this job.

No need to do anything along those lines now, but maybe you could add `TODO(threads)`. If we then get around the whole multi-threaded topic and revisit this code, it's easier to spot such issues.

Reply:
They dont need to be protected by the mutex itself necessarily, since if we have a `&mut GdCell` then we know that calling `is_currently_bound()` on it cannot race with another value. Since we have a `&mut` reference to the cell.

Reply:
I'm not following here; how do we get a `&mut GdCell` if we have a `Pin<GdCell>` at the call site?
(since it doesn't implement `Unpin`)

Reply:
Well more accurately we'd have a `&mut Pin<Box<GdCell>>`, same applies to that.